src/garage/tf/algos/npo2.py

```diff
@@ -0,0 +1,575 @@
+            baseline_var=baseline_var,
+            valid_var=valid_var,
+            policy_state_info_vars=policy_state_info_vars,
+            flat=pol_flat,
+            valid=pol_valid,
+        )
+        policy_opt_inputs = graph_inputs(
+            'PolicyOptInputs',
+            obs_var=obs_var,
+            old_obs_var=old_obs_var,
+            action_var=action_var,
+            reward_var=reward_var,
+            baseline_var=baseline_var,
+            valid_var=valid_var,
+            policy_state_info_vars_list=policy_state_info_vars_list,
+        )
+
+        return policy_loss_inputs, policy_opt_inputs
+
+    def _build_policy_loss(self, i):
+        policy_entropy = self._build_entropy_term(i)
+        rewards = i.reward_var
+
+        if self._maximum_entropy:
+            with tf.name_scope('augmented_rewards'):
+                rewards = i.reward_var + self.policy_ent_coeff * policy_entropy
+
+        with tf.name_scope('policy_loss'):
+            adv = compute_advantages(self.discount,
+                                     self.gae_lambda,
+                                     self.max_path_length,
+                                     i.baseline_var,
+                                     rewards,
+                                     name='adv')
+
+            adv_flat = flatten_batch_tensor(adv, name='adv_flat')
+            adv_valid = filter_valids_tensor(adv_flat,
+                                             i.flat.valid_var,
+                                             name='adv_valid')
+
+            if self.policy.recurrent:
+                adv = tf.reshape(adv, [-1, self.max_path_length])
+
+            # Optionally normalize advantages
+            eps = tf.constant(1e-8, dtype=tf.float32)
+            if self.center_adv:
+                if self.policy.recurrent:
+                    adv = center_advs(adv, axes=[0], eps=eps)
+                else:
+                    adv_valid = center_advs(adv_valid, axes=[0], eps=eps)
+
+            if self.positive_adv:
+                if self.policy.recurrent:
+                    adv = positive_advs(adv, eps)
+                else:
+                    adv_valid = positive_advs(adv_valid, eps)
+
+            # Calculate loss function and KL divergence
+            with tf.name_scope('kl'):
+                if self.policy.recurrent:
+                    kl = self.old_policy.distribution.kl_divergence(
+                        self.policy.distribution)
+                    pol_mean_kl = tf.reduce_sum(
+                        kl * i.valid_var) / tf.reduce_sum(i.valid_var)
+                else:
+                    kl = self.old_policy.distribution.kl_divergence(
+                        self.policy.distribution)
+                    pol_mean_kl = tf.reduce_mean(kl)
+
+            # Calculate vanilla loss
+            with tf.name_scope('vanilla_loss'):
+                if self.policy.recurrent:
+                    ll = self.policy.distribution.log_prob(
```

Comment:
i think this is a good opportunity for you to add `log_prob` and `kl_divergence` to `StochasticPolicy2` (or have StochasticPolicy inherit from `tfp.Distribution` -- not sure which is better). it will make reading the algorithms much easier.

perhaps it will be something like 

```python
class Policy(Model):
    pass

class StochasticPolicy(tfp.Distribution, Policy):
    pass


ll = self.policy.log_prob(i.action_var)
lr = tf.exp(self.policy.log_prob(i.action_var) - self.old_policy.log_prob(i.action_var))
# etc...
```
```